src/protocol/attribution/apply_attribution_window.rs

```diff
@@ -0,0 +1,303 @@
+///
+/// # Errors
+/// Fails if sub-protocols fails.
+#[allow(dead_code)]
+async fn apply_attribution_window<F, C, T>(
+    ctx: C,
+    input: &[MCApplyAttributionWindowInputRow<F, T>],
+    attribution_window_seconds: u32,
+) -> Result<impl Iterator<Item = MCApplyAttributionWindowOutputRow<F, T>> + '_, Error>
+where
+    F: Field,
+    C: Context + RandomBits<F, Share = T>,
+    T: Arithmetic<F> + BasicProtocols<C, F>,
+{
+    let mut t_deltas = prefix_sum_time_deltas(&ctx, input).await?;
+
+    let trigger_values =
+        zero_out_expired_trigger_values(&ctx, input, &mut t_deltas, attribution_window_seconds)
+            .await?;
+
+    Ok(input.iter().zip(trigger_values).map(|(x, value)| {
+        MCApplyAttributionWindowOutputRow::new(
+            x.is_trigger_report.clone(),
+            x.helper_bit.clone(),
+            x.breakdown_key.clone(),
+            value,
+        )
+    }))
+}
+
+/// Computes time deltas from each trigger event to its nearest matching source event.
+///
+/// # Errors
+/// Fails if the multiplication fails.
+async fn prefix_sum_time_deltas<F, C, T>(
+    ctx: &C,
+    input: &[MCApplyAttributionWindowInputRow<F, T>],
+) -> Result<Vec<T>, Error>
+where
+    F: Field,
+    C: Context,
+    T: Arithmetic<F> + BasicProtocols<C, F>,
+{
+    let num_rows = input.len();
+
+    // Pre-compute `is_trigger_bit * helper_bit'.
+    // Later in the prefix-sum loop, this vector is updated in each iteration to help
+    // accumulate timedeltas and determine when to stop accumulating.
+    let stop_bit_context = ctx
+        .narrow(&Step::IsTriggerBitTimesHelperBit)
+        .set_total_records(num_rows - 1);
+    let stop_bits = std::iter::once(T::ZERO).chain(
+        try_join_all(input.iter().skip(1).enumerate().map(|(i, x)| {
+            let c = stop_bit_context.clone();
+            let record_id = RecordId::from(i);
+            async move { T::multiply(c, record_id, &x.is_trigger_report, &x.helper_bit).await }
+        }))
+        .await?,
+    );
+
+    // First, create a vector of timedeltas. This vector contains non-zero values only for
+    // rows with `stop_bit` = 1, meaning that the row is a trigger event, and has the same
+    // match key as the event one above.
+    let t_delta_context = ctx
+        .narrow(&Step::InitializeTimeDelta)
+        .set_total_records(num_rows - 1);
+    let mut t_delta = Some(T::ZERO)
+        .into_iter()
+        .chain(
+            try_join_all(
+                zip(input.iter(), input.iter().skip(1))
+                    .zip(stop_bits.clone().skip(1))
+                    .enumerate()
+                    .map(|(i, ((prev, curr), b))| {
+                        let c = t_delta_context.clone();
+                        let record_id = RecordId::from(i);
+                        let delta = curr.timestamp.clone() - &prev.timestamp;
+                        async move { T::multiply(c, record_id, &delta, &b).await }
```

Comment:
I kind of feel like this should be taken care of by the `do_the_binary_tree_thing` function. I'm not sure why you have to do this multiplication.

Reply:
This part is the conversion from timestamps to time deltas. There might be a clever way of doing this without multiplications.